src/bun.js/api/server.zig

```diff
@@ -7179,19 +7203,37 @@ pub fn NewServer(comptime NamespaceType: type, comptime ssl_enabled_: bool, comp
                 const task = ServerAllConnectionsClosedTask.new(.{
                     .globalObject = this.globalThis,
                     // Duplicate the Strong handle so that we can hold two independent strong references to it.
-                    .promise = JSC.JSPromise.Strong{
+                    .promise = .{
                         .strong = JSC.Strong.create(this.all_closed_promise.value(), this.globalThis),
                     },
                     .tracker = JSC.AsyncTaskTracker.init(vm),
                 });
                 event_loop.enqueueTask(JSC.Task.init(task));
             }
-            if (this.pending_requests == 0 and this.listener == null and this.flags.has_js_deinited and !this.hasActiveWebSockets()) {
+            if (this.pending_requests == 0 and
+                this.listener == null and
+                !this.hasActiveWebSockets())
+            {
                 if (this.config.websocket) |*ws| {
                     ws.handler.app = null;
                 }
                 this.unref();
-                this.scheduleDeinit();
+
+                // Detach DevServer. This is needed because there are aggressive
+                // tests that check for DevServer memory soundness. This reveals
+                // a larger problem, that it seems that some objects like Server
+                // should be detachable from their JSValue, so that when the
+                // native handle is done, keeping the JS binding doesn't use
+                // `this.memoryCost()` bytes.
+                if (this.dev_server) |dev| {
+                    dev.deinit();
```

Comment:
set it to null before calling deinit so that there's no risk of side effects from future refactors

```suggestion
                    this.dev_server = null;
                    dev.deinit();
```